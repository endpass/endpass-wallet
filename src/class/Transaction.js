import { ERC20Token, Token } from '@/class';
import { isNumeric } from '@/utils/numbers';
import web3 from '@/utils/web3';
import { BigNumber } from 'bignumber.js';

export default class Transaction {
  constructor({
    data,
    from,
    gasPrice = '90',
    gasLimit = '22000',
    hash,
    input,
    nonce,
    state = 'success',
    timestamp,
    to = '',
    networkId = 1,
    tokenInfo = undefined,
    transactionHash,
    value = '0',
    success,
  }) {
    if (tokenInfo) {
      this.tokenInfo = new Token(tokenInfo);
      this.valueWei = value;
    } else {
      this.value = value;
    }

    this.data = data || input;
    this.from = from;
    this.networkId = networkId;
    this.gasPrice = gasPrice;
    this.gasLimit = gasLimit;
    this.hash = hash || transactionHash;
    this.nonce = nonce;
    this.state = success === false ? 'error' : state;
    this.to = to;
    if (timestamp) {
      this.date = new Date(timestamp * 1000);
    }
  }

  set value(value) {
    this._value = String(value);
  }

  get value() {
    return this._value;
  }

  get valueWei() {
    if (!isNumeric(this._value)) return '0';

    let multiplier = BigNumber('10').pow(18);

    if (this.tokenInfo) {
      multiplier = BigNumber('10').pow(this.tokenInfo.decimals || 0);
    }

    return BigNumber(this._value)
      .times(multiplier)
      .toFixed(0);
  }

  set valueWei(valueWei) {
    if (!isNumeric(valueWei)) return '0';

    let multiplier = BigNumber('10').pow(18);

    if (this.tokenInfo) {
      multiplier = BigNumber('10').pow(this.tokenInfo.decimals || 0);
    }

    this._value = BigNumber(valueWei)
      .div(multiplier)
      .toFixed();
  }

  set gasPrice(price) {
    this._gasPrice = price.toString();
  }

  get gasPrice() {
    return this._gasPrice;
  }

  set to(to) {
    this._to = web3.utils.isAddress(to) ? web3.utils.toChecksumAddress(to) : to;
  }

  get to() {
    return this._to;
  }

  set from(from) {
    this._from = web3.utils.isAddress(from)
      ? web3.utils.toChecksumAddress(from)
      : from;
  }

  get from() {
    return this._from;
  }

  get gasPriceWei() {
    if (!isNumeric(this.gasPrice)) return '0';

    return web3.utils.toWei(this.gasPrice, 'Gwei');
  }

  set gasLimit(limit) {
    this._gasLimit = limit;
  }

  get gasLimit() {
    return this._gasLimit;
  }

  get validTo() {
    const { to } = this;

    if (!to) {
      return undefined;
    }

    if (/^0x/i.test(to)) {
      return to;
    }

    return `0x${to}`;
  }

  get gasCost() {
    return BigNumber(this.gasPriceWei).times(this.gasLimit);
  }

  get token() {
    const token = this.tokenInfo && this.tokenInfo.symbol;
    return token || 'ETH';
  }

  getValidData() {
    let { data } = this;

    if (this.tokenInfo) {
      const erc20 = new ERC20Token(this.tokenInfo.address);
      const contract = erc20.getContract();

      data = contract.methods.transfer(this.validTo, this.valueWei).encodeABI();
    }

    return data;
  }

  async getFullPrice() {
    const estimation = await this.estimateGas();

    return BigNumber(this.gasPriceWei)
      .times(estimation)
      .toFixed();
  }

  async estimateGas() {
    const estimationParams = {
      data: this.getValidData(),
      to: this.validTo,
    };
    const estimatedGas = await web3.eth.estimateGas(estimationParams);

    return estimatedGas;
  }

  getApiObject() {
    this.data = this.getValidData();
    let tnxData = {
      from: this.from,
      to: this.validTo,
      gasPrice: web3.eth.numberToHex(this.gasPriceWei),
      value: web3.eth.numberToHex(this.valueWei),
      gasLimit: web3.eth.numberToHex(this.gasLimit || 0),
      data: this.data,
      nonce: web3.eth.numberToHex(this.nonce),
    };

    if (this.tokenInfo) {
      tnxData = {
        ...tnxData,
        to: this.tokenInfo.address,
        value: '0x0',
      };
    }

    return tnxData;
  }

  getUpGasPrice() {
    return BigNumber(this.gasPrice)
      .plus('10')
      .integerValue(BigNumber.ROUND_CEIL);
  }

  clone() {
    let tnxData = {
      from: this.from,
      to: this.validTo,
      gasPrice: this.gasPrice,
      value: this.value,
      gasLimit: this.gasLimit,
      data: this.data,
      nonce: this.nonce,
      networkId: this.networkId,
      hash: this.hash,
    };

    if (this.tokenInfo) {
      tnxData = {
        ...tnxData,
        tokenInfo: this.tokenInfo,
        to: this.validTo,
        value: this.valueWei,
      };
    }

    return new Transaction(tnxData);
  }

  static isEqual(trx1, trx2) {
    return (
      `${trx1.networkId}-${trx1.hash}` === `${trx2.networkId}-${trx2.hash}`
    );
  }
}
