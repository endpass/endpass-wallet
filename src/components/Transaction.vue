<template lang="html">
  <div
    :class="statusClass"
    class="transaction"
  >
    <div class="transaction-header">
      <account
        :address="txAddress"
        :balance="transaction.value.toString()"
        :currency="symbol"
      >
        <div class="transaction-actions level is-mobile">
          <div class="level-left">
            <div class="level-item">
              <span class="date">{{ transactionDateFromNow }}</span>
            </div>
          </div>
          <div class="level-right">
            <v-spinner
              v-if="isHavePendingRelatedTransaction && transaction.state === 'pending'"
              :label="pendingActionText"
              class="level-item has-text-info actions-loader"
            />
            <template v-else>
              <a
                v-if="transaction.state === 'pending' && !isPublicAccount"
                :disabled="isSyncing"
                class="level-item has-text-info"
                title="Resend"
                role="button"
                data-test="transaction-resend-button"
                @click="resend"
              >
                <span
                  class="icon is-small"
                  v-html="require('@/img/loop.svg')"
                />
                <span class="caption is-hidden-mobile">
                  Resend
                </span>
              </a>
              <a
                v-if="transaction.state === 'pending' && !isPublicAccount"
                :disabled="isSyncing"
                class="level-item has-text-danger"
                title="Cancel"
                role="button"
                data-test="transaction-cancel-button"
                @click="cancel"
              >
                <span
                  class="icon is-small"
                  v-html="require('@/img/ban.svg')"
                />
                <span class="caption is-hidden-mobile">
                  Cancel
                </span>
              </a>
            </template>

            <a
              title="Details"
              class="level-item has-text-info"
              @click="toggleExpanded"
            >
              <span
                class="icon is-small"
                v-html="require('@/img/ellipses.svg')"
              />

              <span class="caption is-hidden-mobile">Details</span>
            </a>
          </div>
        </div>
      </account>
    </div>

    <div
      v-if="isExpanded"
      class="transaction-detail"
    >
      <div v-if="transaction.hash.length">
        <span class="text-label">Txid</span>
        <p class="code">{{ transaction.hash }}</p>
      </div>
      <div>
        <span class="heading status-text">{{ transaction.state }}</span>
      </div>

      <div v-if="transaction.date">
        <span class="text-label">Date</span>
        <p class="date">{{ transactionFormatedDate }}</p>
      </div>

      <div v-if="recieve">
        <span class="text-label">From</span>
        <p class="code address">{{ transaction.from }}</p>
      </div>

      <div v-else>
        <span class="text-label">To</span>
        <p class="code address">{{ transaction.to }}</p>
      </div>

      <div>
        <span class="text-label">Nonce</span>
        <span class="">{{ transaction.nonce }}</span>
      </div>

      <div
        v-if="parsedData"
      >
        <span class="text-label">Data</span>
        <p class="code">
          {{ parsedData }}
        </p>
      </div>
    </div>

    <resend-modal
      v-if="resendModalOpen"
      :transaction="transactionToSend"
      @close="closeResendModal"
      @confirm="confirmResend"
    />
    <password-modal
      v-if="passwordModalOpen"
      @close="closePassword"
      @confirm="confirmPassword"
    />
  </div>
</template>

<script>
import dayjs from 'dayjs';
import Account from '@/components/Account';
import ResendModal from '@/components/modal/ResendModal';
import PasswordModal from '@/components/modal/PasswordModal';
import VSpinner from '@/components/ui/VSpinner';
import { mapState, mapGetters, mapActions } from 'vuex';
import { fromNow, formateDate } from '@/utils/date';
import { getShortStringWithEllipsis } from '@/utils/strings';
import web3 from '@/class/singleton/web3';

const { hexToString } = web3.utils;

export default {
  props: {
    transaction: {
      type: Object,
      required: true,
    },
  },

  data() {
    return {
      resendModalOpen: false,
      passwordModalOpen: false,
      isExpanded: false, // details are expanded
      transactionToSend: null,
      state: null,
      displayDate: this.transaction.date,
      dateTimer: null,
    };
  },
  computed: {
    ...mapState({
      address: state => state.accounts.address,
      isSyncing: state => !!state.connectionStatus.isSyncing,
    }),
    ...mapGetters('accounts', ['isPublicAccount']),
    recieve() {
      return this.transaction.to === this.address;
    },
    isSuccess() {
      return this.transaction.state === 'success';
    },
    isError() {
      return (
        this.transaction.state === 'error' ||
        this.state === 'error' ||
        this.state === 'canceled'
      );
    },
    isPending() {
      return this.transaction.state === 'pending';
    },
    pendingActionText() {
      let text = '';

      if (this.isHavePendingRelatedTransaction) {
        if (this.state === 'canceled') {
          text = 'Canceling...';
        } else if (this.state === 'resent') {
          text = 'Resending...';
        }
      }

      return text;
    },
    isHavePendingRelatedTransaction() {
      return (
        this.transactionToSend && this.transactionToSend.state === 'pending'
      );
    },
    // Dynamic class based on transaction status
    statusClass() {
      return {
        'is-success': this.isSuccess,
        'is-warning': this.isPending,
        'is-danger': this.isError,
        'is-received': this.recieve,
        'is-sent': !this.recieve,
      };
    },
    symbol() {
      return (
        (this.transaction.tokenInfo && this.transaction.tokenInfo.symbol) ||
        'ETH'
      );
    },

    transactionFormatedDate() {
      return formateDate(this.transaction.date);
    },

    transactionDateFromNow() {
      return fromNow(this.displayDate);
    },

    txAddress() {
      if (this.recieve) {
        return this.transaction.from;
      }
      return this.transaction.to;
    },

    parsedData() {
      const dataString = this.transaction.data || '0x';

      return hexToString(dataString);
    },
  },

  methods: {
    ...mapActions('transactions', ['resendTransaction', 'cancelTransaction']),
    requestPassword() {
      this.passwordModalOpen = true;
    },
    closePassword() {
      this.passwordModalOpen = false;
    },
    confirmPassword(password) {
      this.resendModalOpen = false;
      this.passwordModalOpen = false;

      const sendTransaction =
        this.state === 'canceled'
          ? this.cancelTransaction
          : this.resendTransaction;

      this.transactionToSend.state = 'pending';
      sendTransaction({ transaction: this.transactionToSend, password })
        .then(() => {
          this.$notify({
            title: 'Successful',
            text: `Transaction was ${this.state}`,
            type: 'is-info',
          });
          this.transactionToSend = null;
          this.state = null;
        })
        .catch(() => {});
    },
    confirmResend(transaction) {
      this.transactionToSend = transaction;
      this.requestPassword();
    },
    toggleExpanded() {
      this.isExpanded = !this.isExpanded;
    },
    closeResendModal() {
      this.resendModalOpen = false;
    },
    resend() {
      this.transactionToSend = this.transaction.clone();
      this.transactionToSend.state = null;
      this.resendModalOpen = true;
      this.state = 'resent';
    },
    cancel() {
      if (this.transaction.state !== 'pending') return;

      this.state = 'canceled';
      this.transactionToSend = this.transaction.clone();
      this.transactionToSend.value = '0';
      this.transactionToSend.to = this.address;
      this.transactionToSend.gasPrice = this.transactionToSend.getUpGasPrice();
      this.transactionToSend.state = null;
      this.requestPassword();
    },
    incrementDIsplayDate() {
      this.displayDate = dayjs(this.transaction.date)
        .add(10, 's')
        .toDate();
    },
  },

  created() {
    this.dateTimer = setInterval(this.incrementDIsplayDate, 10000);
  },

  beforeDestroy() {
    clearInterval(this.dateTimer);
    this.dateTimer = null;
  },

  components: {
    Account,
    ResendModal,
    PasswordModal,
    VSpinner,
  },
  filters: {
    truncateHash(value) {
      if (!value) return '';

      return getShortStringWithEllipsis(value.toString());
    },
  },
};
</script>

<style lang="scss">
.transaction {
  margin-bottom: 2rem;
  .card {
    border-top: 5px solid $white;
    overflow: hidden;
    text-overflow: ellipsis;
  }
  &.is-danger .card {
    border-top-color: $danger;
    .status-text {
      color: $danger;
    }
  }
  &.is-warning .card {
    border-top-color: $warning;
    .status-text {
      color: $warning;
    }
  }
  &.is-success .card {
    border-top-color: $success;
    .status-text {
      color: $success;
    }
  }
  &.is-received {
    .balance::before {
      content: '+';
      color: $success;
    }
    .status-icon svg {
      fill: $success;
    }
    .amount {
      color: $success;
    }
  }
  &.is-sent {
    .balance::before {
      content: '-';
      color: $danger;
    }
    .status-icon svg {
      fill: $danger;
    }
    .amount {
      color: $danger;
    }
  }

  .account {
    .media-right {
      margin-left: 0;
    }
    .balance {
      line-height: normal;
    }
    .amount {
      font-size: 1.2em;
    }
    .currency {
      font-size: 0.8em;
    }
  }

  .date {
    color: $dark-grey;
    font-size: 0.9em;
    font-style: italic;
  }

  .card-header {
    flex-direction: column;
    .card-header-link {
      cursor: pointer;
      &:hover {
        background-color: rgba(75, 4, 114, 0.2);
      }
    }
  }

  .text-label {
    color: lighten($dark-grey, 20%);
    margin-right: 0.2em;
  }

  .card-header-title {
    * {
      margin-left: 0.5rem;
    }
  }

  .actions-loader .sl-spinner {
    width: 16px !important;
    height: 16px !important;
  }
}
</style>
