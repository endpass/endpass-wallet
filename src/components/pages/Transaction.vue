<template>
  <base-page>
    <template slot="title">
      Transaction
    </template>
    <div class="tabs">
      <ul>
        <li
          v-for="tab in tabs"
          :key="tab.id"
          :class="{ 'is-active': isTabActive(tab.id) }"
        >
          <a
            role="button"
            @click="switchTab(tab.id)"
          >
            {{ tab.label }}
          </a>
        </li>
      </ul>
    </div>
    <sign-transaction v-show="currentTab === 'sign'" />
    <recover-transaction v-show="currentTab === 'recover'" />
  </base-page>
</template>

<script>
import BasePage from '@/components/pages/Base';
import SignTransaction from '@/components/SignTransaction';
import RecoverTransaction from '@/components/RecoverTransaction';
import privatePage from '@/mixins/privatePage';

export default {
  name: 'TransactionPage',

  data: () => ({
    currentTab: 'sign',
    tabs: [
      {
        label: 'Sign',
        id: 'sign',
      },
      {
        label: 'Recover',
        id: 'recover',
      },
    ],
  }),

  methods: {
    switchTab(tab) {
      this.currentTab = tab;
    },

    isTabActive(tab) {
      return this.currentTab === tab;
    },
  },

  mixins: [privatePage],

  components: {
    BasePage,
    SignTransaction,
    RecoverTransaction,
  },
};
</script>
