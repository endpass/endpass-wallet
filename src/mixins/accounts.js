// Mixin for common methods for Ethereum accounts
import web3 from 'web3';

const { fromWei } = web3.utils;

export default {
  computed: {
    address() {
      return this.$store.state.accounts.address;
    },

    balance() {
      return this.$store.state.accounts.balance === null
        ? null
        : fromWei(this.$store.getters['accounts/balance']);
    },
  },
};
