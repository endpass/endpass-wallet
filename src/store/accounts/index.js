import actions from './actions';
import getters from './getters';
import mutations from './mutations';

/**
 * @namespace
 * @property {Boolean} isInited Status of initializing accounts module
 * @property {Object} hdKey The encrypted xprv key for the user's hd wallet as a V3 keystore object.
 *  All accounts are direct child accounts.
 * @property {Object} hadrwareXpub The user's hardware xpub, keyed by wallet type
 * @property {Object} wallets The user's wallets, mapped by address
 * @property {String} address Current wallet address
 * @property {String} balance Currect account balance
 */
const state = {
  isInited: false,
  hardwareXpub: {},
  wallets: {},
  hdKey: null,
  address: null,
  balance: null,
};

export default {
  namespaced: true,
  state,
  actions,
  getters,
  mutations,
};
