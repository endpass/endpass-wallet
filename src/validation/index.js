import Vue from 'vue';
import { default as VeeValidate, Validator } from 'vee-validate';
import address from './address';
import privateKey from './privateKey';
import publicKey from './publicKey';
import seedPhrase from './seedPhrase';
import ensOrAddress from './ensOrAddress';
import hex from './hex';

Validator.extend('address', address);
Validator.extend('ensOrAddress', ensOrAddress, {
  immediate: false,
});
Validator.extend('private_key', privateKey);
Validator.extend('public_key', publicKey);
Validator.extend('seed_phrase', seedPhrase);
Validator.extend('hex', hex);

Vue.use(VeeValidate, { inject: false });
