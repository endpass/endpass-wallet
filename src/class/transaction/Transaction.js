<<<<<<< HEAD
import { cloneDeep, get } from 'lodash';
import { BigNumber } from 'bignumber.js';
import { isNumeric } from '@endpass/utils/numbers';
import { ERC20Token } from '@/class';
import web3 from '@/class/singleton/web3';

import applyProps from './applyProps';

const { toWei, numberToHex, isAddress } = web3.utils;
=======
import cloneDeep from 'lodash/cloneDeep';
import get from 'lodash/get';
import { BigNumber } from 'bignumber.js';
import { isNumeric } from '@endpass/utils/numbers';
import { toWei, numberToHex, isAddress } from 'web3-utils';
import { ERC20Token } from '../internal';

import applyProps from './applyProps';
import injectWeb3 from '../injectWeb3';

let web3;
>>>>>>> refactor: Step first. Move EventEmitter to core/class

const DEFAULT_ZERO = '0';

export default class Transaction {
<<<<<<< HEAD
=======
  static set web3(val) {
    web3 = val;
  }

>>>>>>> refactor: Step first. Move EventEmitter to core/class
  static create({
    data,
    from,
    gasPrice = '90',
    gasLimit = '22000',
    hash,
    input,
    nonce,
    state = 'success',
    timestamp,
    to = '',
    networkId = 1,
    tokenInfo = undefined,
    transactionHash,
    value = DEFAULT_ZERO,
    success,
  }) {
    // prepare external structure
    const trx = {
      tokenInfo,
      from,
      to,
      networkId,
      gasPrice,
      gasLimit,
      nonce,
      // dependency from other props
      data: data || input,
      hash: hash || transactionHash,
      state: success === false ? 'error' : state,
      date: timestamp && new Date(timestamp * 1000),
      // calculated
      value: DEFAULT_ZERO,
    };

    applyProps(trx, {
      value,
      from,
      gasPrice,
      gasLimit,
      to,
    });

    return trx;
  }

  static applyProps(trx, newProps) {
<<<<<<< HEAD
    const res = Transaction.unfreeze(trx);
=======
    const res = { ...trx }; // unfreeze
>>>>>>> refactor: Step first. Move EventEmitter to core/class
    applyProps(res, newProps);
    return res;
  }

<<<<<<< HEAD
  static unfreeze(trx) {
    return { ...trx };
  }

=======
>>>>>>> refactor: Step first. Move EventEmitter to core/class
  static clone(trx) {
    return cloneDeep(trx);
  }

  static isEqual(trx1, trx2) {
    return (
      `${trx1.networkId}-${trx1.hash}` === `${trx2.networkId}-${trx2.hash}`
    );
  }

  static async isToContract(transaction) {
    const res = await web3.eth.getCode(transaction.to);

    return res !== '0x';
  }

  static getMultiplier(tokenInfo) {
    const decimals = tokenInfo ? tokenInfo.decimals || 0 : 18;
    return BigNumber('10').pow(decimals);
  }

  static getTokenSymbol(trx) {
    const { tokenInfo } = trx;
    return get(tokenInfo, 'symbol', 'ETH');
  }

  static getGasCost({ gasLimit, gasPrice }) {
    const gasPriceWei = Transaction.getGasPriceWei(gasPrice);
    return BigNumber(gasPriceWei).times(gasLimit);
  }

  static getUpGasCost(trx) {
    const { token } = trx;
    const gasCost = Transaction.getGasCost(trx);

    const tnxValue =
      token === 'ETH' ? Transaction.getValueInWei(trx) : DEFAULT_ZERO;

    return BigNumber(gasCost).plus(tnxValue);
  }

  static getUpGasPrice({ gasPrice }) {
    return BigNumber(gasPrice)
      .plus('10')
      .integerValue(BigNumber.ROUND_CEIL);
  }

  static getGasPriceWei(price) {
    const value = price && price.toString();
    return !isNumeric(value) ? DEFAULT_ZERO : toWei(value, 'Gwei');
  }

  static getValidTo(transaction) {
    const { to } = transaction;

    if (!to) {
      return undefined;
    }

    if (isAddress(to)) {
      return to;
    }

    return `0x${to}`;
  }

  static getValidData(transaction) {
    const { data, tokenInfo } = transaction;
    if (!tokenInfo) {
      return data;
    }

    const validTo = Transaction.getValidTo(transaction);
    const erc20 = new ERC20Token(tokenInfo.address);
    const contract = erc20.getContract();
    const transactionValueInWei = Transaction.getValueInWei(transaction);
    return contract.methods
      .transfer(validTo, transactionValueInWei)
      .encodeABI();
  }

  static getPriceWei(transaction) {
    return Transaction.getGasPriceWei(transaction.gasPrice);
  }

  static getValueBN(transaction) {
    const { value } = transaction;

    if (!isNumeric(value)) {
      return BigNumber(DEFAULT_ZERO);
    }

    return BigNumber(value);
  }

  static getValueInWei(transaction) {
    const multiplier = Transaction.getMultiplier(transaction.tokenInfo);
    return Transaction.getValueBN(transaction)
      .times(multiplier)
      .toFixed(0);
  }

  static getValueFromWei(transaction) {
    const multiplier = Transaction.getMultiplier(transaction.tokenInfo);
    return Transaction.getValueBN(transaction)
      .div(multiplier)
      .toFixed();
  }

  static async getGasFullPrice(transaction) {
<<<<<<< HEAD
    const estimationParams = {
      data: Transaction.getValidData(transaction),
      to: Transaction.getValidTo(transaction),
    };
    const estimatedGas = await web3.eth.estimateGas(estimationParams);
=======
    const estimatedGas = await web3.eth.estimateGas({
      data: Transaction.getValidData(transaction),
      to: Transaction.getValidTo(transaction),
    });
>>>>>>> refactor: Step first. Move EventEmitter to core/class
    const gasPriceWei = Transaction.getPriceWei(transaction);

    return BigNumber(gasPriceWei)
      .times(estimatedGas)
      .toFixed();
  }

  static getApiObject(trx) {
    const data = Transaction.getValidData(trx);
    const validTo = Transaction.getValidTo(trx);
    const gasPriceWei = Transaction.getPriceWei(trx);
    const valueWei = Transaction.getValueInWei(trx);
    const { tokenInfo } = trx;

    const tokenPassData = tokenInfo
      ? {
          to: tokenInfo.address,
          value: '0x0',
        }
      : {};

    return {
      from: trx.from,
      to: validTo,
      gasPrice: numberToHex(gasPriceWei),
      value: numberToHex(valueWei),
      gasLimit: numberToHex(trx.gasLimit || 0),
      nonce: numberToHex(trx.nonce),
      data,
      ...tokenPassData,
    };
  }
<<<<<<< HEAD

  static async getFullPrice(trx) {
    const estimation = await Transaction.estimateGas(trx);
    const gasPriceWei = Transaction.getPriceWei(trx);

    return BigNumber(gasPriceWei)
      .times(estimation)
      .toFixed();
  }

  static async estimateGas(trx) {
    const validTo = Transaction.getValidTo(trx);
    const estimationParams = {
      data: Transaction.getValidData(trx),
      to: validTo,
    };
    const estimatedGas = await web3.eth.estimateGas(estimationParams);

    return estimatedGas;
  }
}
=======
}

const createTransactionClass = injectWeb3(Transaction);

export { createTransactionClass };
>>>>>>> refactor: Step first. Move EventEmitter to core/class
