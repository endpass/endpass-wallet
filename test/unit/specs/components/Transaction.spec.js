import Vuex from 'vuex';
import Notifications from 'vue-notification';
import VeeValidate from 'vee-validate';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import UIComponents from '@endpass/ui';
import validation from '@/validation';

import Transaction from '@/components/Transaction';

import { checksumAddress } from 'fixtures/accounts';

const localVue = createLocalVue();

localVue.use(Notifications);
localVue.use(Vuex);
localVue.use(validation);
localVue.use(VeeValidate);
localVue.use(UIComponents);

describe('Transaction', () => {
  const accountsActions = {
    resendTransaction: jest.fn(),
    cancelTransaction: jest.fn(),
  };
  let wrapper;

  beforeEach(() => {
    const storeOptions = {
      modules: {
        accounts: {
          namespaced: true,
          state: {
            address: checksumAddress,
          },
          getters: {
            isPublicAccount: jest.fn(() => false),
          },
          accountsActions,
        },
        connectionStatus: {
          namespaced: true,
          state: {
            isSyncing: false,
          },
        },
      },
    };
    const transaction = {
      state: 'pending',
      value: 0,
      from: 'fromAddr',
      to: 'toAddr',
    };
    const store = new Vuex.Store(storeOptions);

    wrapper = shallowMount(Transaction, {
      localVue,
      store,
      propsData: {
        transaction,
      },
    });
  });

  afterEach(() => {
    accountsActions.resendTransaction.mockClear();
    accountsActions.cancelTransaction.mockClear();
  });

  describe('render', () => {
    it('should render the initial state of the component', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    describe('action process', () => {
      it('should render loader with pending text', () => {
        wrapper.setData({
          state: 'resent',
          transactionToSend: {
            state: 'pending',
          },
        });

        expect(wrapper.find('v-spinner-stub').html()).toMatchSnapshot();
      });

      it('should render loader with canceling text', () => {
        wrapper.setData({
          state: 'canceled',
          transactionToSend: {
            state: 'pending',
          },
        });

        expect(wrapper.find('v-spinner-stub').html()).toMatchSnapshot();
      });

      it('should not render loader', () => {
        wrapper.setData({
          transactionToSend: null,
        });

        expect(wrapper.find('v-spinner-stub').exists()).toBeFalsy();
      });
    });
  });

  describe('behavior', () => {
    beforeEach(() => {
      jest.useFakeTimers();
    });

    it('should start timer at component creating', () => {
      expect(wrapper.vm.dateTimer).not.toBeNull();
    });

    it('should add one minute each minute to display date', () => {
      const date = new Date();

      wrapper.setData({
        displayDate: date,
      });

      jest.runOnlyPendingTimers();
      jest.advanceTimersByTime(10000);

      expect(wrapper.vm.displayDate).not.toBe(date);
      const displaceDate = new Date(wrapper.vm.displayDate);
      const diff = Math.abs(date - displaceDate);

      expect(Math.floor(diff / 1000)).toBe(10); // 10 seconds
    });
  });
});
