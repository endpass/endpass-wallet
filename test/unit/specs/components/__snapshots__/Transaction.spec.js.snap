// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transaction render action process should render loader 1`] = `<v-spinner-stub label="Resending..." class="level-item has-text-info actions-loader"></v-spinner-stub>`;

exports[`Transaction render action process should render loader 2`] = `<v-spinner-stub label="Resending..." class="level-item has-text-info actions-loader"></v-spinner-stub>`;

exports[`Transaction render should render the initial state of the component 1`] = `
<div
  class="transaction is-warning is-sent"
>
  <div
    class="transaction-header"
  >
    <account-stub
      address="toAddr"
      balance="0"
      currency="ETH"
      size="50"
    >
      <div
        class="transaction-actions level is-mobile"
      >
        <div
          class="level-left"
        >
          <div
            class="level-item"
          >
            <span
              class="date"
            >
              a few seconds ago
            </span>
          </div>
        </div>
         
        <div
          class="level-right"
        >
          <a
            class="level-item has-text-info"
            data-test="transaction-resend-button"
            role="button"
            title="Resend"
          >
            <span
              class="icon is-small"
            >
              {}
            </span>
             
            <span
              class="caption is-hidden-mobile"
            >
              
                Resend
              
            </span>
          </a>
           
          <a
            class="level-item has-text-danger"
            data-test="transaction-cancel-button"
            role="button"
            title="Cancel"
          >
            <span
              class="icon is-small"
            >
              {}
            </span>
             
            <span
              class="caption is-hidden-mobile"
            >
              
                Cancel
              
            </span>
          </a>
           
          <a
            class="level-item has-text-info"
            data-test="transaction-details-button"
            title="Details"
          >
            <span
              class="icon is-small"
            >
              {}
            </span>
             
            <span
              class="caption is-hidden-mobile"
            >
              Details
            </span>
          </a>
        </div>
      </div>
    </account-stub>
  </div>
   
  <!---->
   
  <!---->
   
  <!---->
</div>
`;
